src/coreclr/jit/importer.cpp

```diff
@@ -2869,54 +2869,17 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
                     return 0;
                 }
 
-                GenTree* const treeToBox       = impStackTop().val;
-                bool           canOptimize     = true;
-                GenTree*       treeToNullcheck = nullptr;
-
-                // Can the thing being boxed cause a side effect?
-                if ((treeToBox->gtFlags & GTF_SIDE_EFFECT) != 0)
-                {
-                    // Is this a side effect we can replicate cheaply?
-                    if (((treeToBox->gtFlags & GTF_SIDE_EFFECT) == GTF_EXCEPT) && treeToBox->OperIs(GT_BLK, GT_IND))
-                    {
-                        // If the only side effect comes from the dereference itself, yes.
-                        GenTree* const addr = treeToBox->AsOp()->gtGetOp1();
-
-                        if ((addr->gtFlags & GTF_SIDE_EFFECT) != 0)
-                        {
-                            canOptimize = false;
-                        }
-                        else if (fgAddrCouldBeNull(addr))
-                        {
-                            treeToNullcheck = addr;
-                        }
-                    }
-                    else
-                    {
-                        canOptimize = false;
-                    }
-                }
-
-                if (canOptimize)
+                if ((opts == BoxPatterns::IsByRefLike) ||
+                    (info.compCompHnd->getBoxHelper(pResolvedToken->hClass) == CORINFO_HELP_BOX))
                 {
-                    if ((opts == BoxPatterns::IsByRefLike) ||
-                        info.compCompHnd->getBoxHelper(pResolvedToken->hClass) == CORINFO_HELP_BOX)
+                    GenTree* op = impPopStack().val;
+                    if ((op->gtFlags & GTF_SIDE_EFFECT) != 0)
                     {
-                        JITDUMP("\n Importing BOX; BR_TRUE/FALSE as %sconstant\n",
-                                treeToNullcheck == nullptr ? "" : "nullcheck+");
-                        impPopStack();
-
-                        GenTree* result = gtNewIconNode(1);
-
-                        if (treeToNullcheck != nullptr)
-                        {
-                            GenTree* nullcheck = gtNewNullCheck(treeToNullcheck, compCurBB);
-                            result             = gtNewOperNode(GT_COMMA, TYP_INT, nullcheck, result);
-                        }
-
-                        impPushOnStack(result, typeInfo(TYP_INT));
-                        return 0;
+                        impStoreTemp(lvaGrabTemp(true DEBUGARG("spill side effects")), op, CHECK_SPILL_ALL);
                     }
```

Comment:
This is a bit of an unusual way of doing this in the importer, use `impSpillSideEffect` instead?

Also we already do this from the caller for `BoxPatterns::IsByRefLike`. Should it be unified?

Reply:
Ah, thanks for the hint. addressed.

> Also we already do this from the caller for BoxPatterns::IsByRefLike. Should it be unified?

Looks like that one has to be always done regardless wether we match a pattern or not while here we only do this when we match it. Presumably the byref case is rare so it's ok to call it twice?

Reply:
> Looks like that one has to be always done regardless wether we match a pattern or not while here we only do this when we match it.

I don't think so (note that the comment is misleading -- we `BADCODE` today on unmatched patterns, we do not generate code to throw `InvalidProgramException` at runtime). But I'm ok with leaving it to avoid further risk and potentially cleaning it up in .NET 9, assuming you want this in .NET 8.

Reply:
Ah, I see, yeah let's clean it up in .NET 9.0, it has a potential to slightly refactor the whole thing